flang/lib/Optimizer/Transforms/ControlFlowConverter.cpp

```diff
@@ -72,6 +75,22 @@ class CfgLoopConv : public mlir::OpRewritePattern<fir::DoLoopOp> {
         rewriter.splitBlock(conditionalBlock, conditionalBlock->begin());
     auto *lastBlock = &loop.getRegion().back();
 
+    // Insert stacksave/stackrestore if there is fir.alloca operation in the
+    // loop.
+    if (hasAllocas) {
+      auto mod = loop.getOperation()->getParentOfType<mlir::ModuleOp>();
+      fir::FirOpBuilder builder(rewriter, mod);
+      builder.setInsertionPointToStart(firstBlock);
+      mlir::func::FuncOp stackSave = fir::factory::getLlvmStackSave(builder);
```

Comment:
This seems to require making the pass a module pass.  This is quite unfortunate.

We can try to use LLVM dialect `stacksave/stackrestore` intrinsics operations, but they do not look too reliable to me, e.g. they do not have any side effects that prevent moving them around `fir.alloca`. At the same time, `fir.call` seems to have the same problem.

Would that be possible to insert stacksave/stackrestore calls during `hlfir.concat` bufferization?  It might be inefficient, but we will get correct code at least.  The redundant stacksave/stackrestore might be optimized after that.

Reply:
> This seems to require making the pass a module pass. This is quite unfortunate.
Yeah if we can avoid that it would probably be better. 

> Would that be possible to insert stacksave/stackrestore calls during hlfir.concat bufferization? It might be inefficient, but we will get correct code at least. The redundant stacksave/stackrestore might be optimized after that.

Yes we can be then we would miss other case (if any) of fir.alloca with dynamic size in loops.

Reply:
You are right, Jean.